src/include/denc.h

```diff
@@ -882,7 +882,12 @@ namespace _denc {
     static void bound_encode(const container& s, size_t& p, uint64_t f = 0) {
       p += sizeof(uint32_t);
       if constexpr (traits::bounded) {
-        if (!s.empty()) {
+        const auto elem_num = s.size();
+        // intensionally not calling container's empty() method to not prohibit
+        // compiler from optimizing the check if it and the ::size() operate on
+        // different memory (observed when std::list::empty() works on pointers,
+        // not the size field).
```

Comment:
@rzarzynski i am wondering if you are optimizing the case where

1. `C<Ts...>::size()` is O(1)
1. if it's not O(1), the majority `std::list` we are encoding is not empty.

but the `std::list::size()` GCC-7 offered by SCLs is O(n). so i can hardly tell if this change is a pure win.

Reply:
Yeah, this change targets solely the 1st case. For C++11-incompliant implementations of `std::list::empty` a penalty can be expected. The question is for how long we need to support those cases. For sure backporting the commit isn't an option but maybe the next release is free of the legacy?

Reply:
GCC-8 from SCLs is still using the C++11-incompliant ABI.

Reply:
if we are able to detect this at build time, probably we can conditionalize on this. the ugliest change i can think of is to add a cmake option to do this. for instance, pass `-DCXX11_COMPATIBLE=OFF` only on centos/rhel7.

i can help on this if you think it's a viable way.

Reply:
Thanks for the commit, @tchaikov!